Nullable.md

```diff
@@ -0,0 +1,442 @@
+the use of `Some{T}` instead of `T` provides a way of having the `T` and `?T`
+types co-exist without specificity issues, but it simultaneously ensures
+that no existing methods for `T` automatically apply to `?T`. It provides
+arguably cleaner semantics (by making `T` and `?T` disjoint) at the cost of
+making the non-null case for nullable types less convenient to work with.
+
+As such, this implementation of `?T` would require that we push users towards
+the consistent usage of specialized lifting machinery that knows how to handle
+`f(x::Union{Some{T}, Void})` given only knowledge of how `f(x::T)` works. For
+one approach to this problem, see the later section on lifting functions from
+non-nullable types to nullable types.
+
+Because users will have to work with lifted functions, this approach makes it
+more difficult for users to write code that works for the non-null case, but
+does not work for the null case. But the possibility of defining methods for
+only one of `Some{T}` or `Void` means that a totality checker would still be
+needed because users can always define only one of `f(x::Some{T})` and
+`f(x::Void)`. (We should note that the current implementation of `Nullable{T}`
+also does not ensure totality.)
+
+Another point in favor of `Some{T}` is that it is sometimes useful to
+distinguish between two confusable outcomes in cases where a method returns
+something like `Union{Some{Void}, Void}` -- as might happen in a hash table
+implementation that returns `Some(nothing)` when a key is present and
+associated with `nothing` as a value, but returns `nothing` when a key is not
+present. This is another implication of working with disjoint types for `T` and
+`?T`.
+
+On the other hand, the use of `Some{T}` raises issues about how we expect `T`,
+`Some{T}` and `Void` to interact. For example, consider the following code:
+
+```
+function sum(xs::Vector{Union{Some{Int}, Void}})
+    s = 0
+    for x in xs
+        s += x
+    end
+    return x
+end
+```
+
+What type does this return when `xs = ?Int[]`? What type does it return when
+`xs = ?Int[Some(1)]`? What type does it return when `xs = ?Int[nothing]`?
+
+At a minimum, the use of `Some{T}` would require careful thought about
+type conversion and type promotion rules since we would need to define methods
+for combinations like `+(x::Int, y::Some{T})`. Almost surely, we would need to
+allow promoting `T` to `Some{T}` to make common code patterns work without
+much verbosity.
+
+### `Union{T, Null{T}}` Pros & Cons
+
+This final approach allows us to distinguish `Null{Bool}` from `Null{Int}` in
+cases in which that distinction might be useful. R makes such a distinction,
+although it seems to be done only because it is falls out naturally from an
+implementation that uses distinct sentinel values for each nullable type in
+the language. Otherwise, this approach is similar to the `Union{T, Void}` case
+described above, except for one crucial problem that would seem to suggest this
```

Comment:
For balance it would be fair to also state it's one "crucial" advantage:

`Union{T1, Null{T1}}` and `Union{T2, Null{T2}}` are disjoint, removing one major class of ambiguity problem for `f(x::?T1)` and `f(x::?T2)` which exists for the other two options.

(actually I just realized that it could be even better if `Null{T}` could be covariant in `T` - that also helps in the case that inference for `return_type` is too wide... speculatively - maybe covariance is helpful for `Some{T}` too?).

A final 4th class of nullable would be `Union{Some{T}, Null{T}}` - has that been discussed?